File: lib/limousine/limousine_transfer_endpoint.dart
// Additive typed itinerary endpoint for limousine quote/book payloads.
// `from`/`to` stay backward-compatible display strings. This snapshot is
// immutable after submit: later provider result changes do not rewrite it.

import '../airport/airport_catalog_repository.dart';
import 'limousine_address_lookup.dart';

abstract final class LimousineTransferEndpointKind {
  static const String address = 'address';
  static const String airport = 'airport';
  static const String hotel = 'hotel';
  static const String event = 'event';
  static const String venue = 'venue';

  static const List<String> all = <String>[
    address,
    airport,
    hotel,
    event,
    venue,
  ];

  static bool isEvent(String kind) => kind == event || kind == venue;
}

class LimousineTransferEndpoint {
  const LimousineTransferEndpoint({
    this.kind = LimousineTransferEndpointKind.address,
    this.displayName = '',
    this.formattedAddress = '',
    this.latitude,
    this.longitude,
    this.providerPlaceId,
    this.airportName,
    this.iataCode,
    this.countryCode,
    this.hotelName,
    this.venueName,
    this.eventName,
    this.city,
    this.postcode,
    this.ratehawkHotelId,
    this.manual = false,
  });

  final String kind;
  final String displayName;
  final String formattedAddress;
  final double? latitude;
  final double? longitude;
  final String? providerPlaceId;
  final String? airportName;
  final String? iataCode;
  final String? countryCode;
  final String? hotelName;
  final String? venueName;
  final String? eventName;
  final String? city;
  final String? postcode;
  final String? ratehawkHotelId;
  final bool manual;

  bool get hasCoordinates =>
      latitude != null &&
      longitude != null &&
      limousineCoordinatesAreValid(latitude!, longitude!);

  String get routeText {
    final formatted = formattedAddress.trim();
    if (formatted.isNotEmpty) return formatted;
    return displayName.trim();
  }

  bool get isEmpty => routeText.isEmpty;

  LimousineTransferEndpoint copyWith({
    String? kind,
    String? displayName,
    String? formattedAddress,
    double? latitude,
    double? longitude,
    String? providerPlaceId,
    String? airportName,
    String? iataCode,
    String? countryCode,
    String? hotelName,
    String? venueName,
    String? eventName,
    String? city,
    String? postcode,
    String? ratehawkHotelId,
    bool? manual,
    bool clearAirport = false,
    bool clearHotel = false,
    bool clearEvent = false,
    bool clearEventName = false,
    bool clearPlaceId = false,
  }) {
    return LimousineTransferEndpoint(
      kind: kind ?? this.kind,
      displayName: displayName ?? this.displayName,
      formattedAddress: formattedAddress ?? this.formattedAddress,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      providerPlaceId: clearPlaceId ? null : (providerPlaceId ?? this.providerPlaceId),
      airportName: clearAirport ? null : (airportName ?? this.airportName),
      iataCode: clearAirport ? null : (iataCode ?? this.iataCode),
      countryCode: countryCode ?? (clearAirport ? this.countryCode : this.countryCode),
      hotelName: clearHotel ? null : (hotelName ?? this.hotelName),
      venueName: clearEvent ? null : (venueName ?? this.venueName),
      eventName: clearEvent || clearEventName ? null : (eventName ?? this.eventName),
      city: city ?? this.city,
      postcode: postcode ?? this.postcode,
      ratehawkHotelId: clearHotel ? null : (ratehawkHotelId ?? this.ratehawkHotelId),
      manual: manual ?? this.manual,
    );
  }

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      'kind': kind,
      'display_name': displayName.trim(),
      'formatted_address': formattedAddress.trim(),
      if (latitude != null) 'latitude': latitude,
      if (longitude != null) 'longitude': longitude,
      if ((providerPlaceId ?? '').trim().isNotEmpty)
        'provider_place_id': providerPlaceId!.trim(),
      if ((airportName ?? '').trim().isNotEmpty) 'airport_name': airportName!.trim(),
      if ((iataCode ?? '').trim().isNotEmpty) 'iata_code': iataCode!.trim().toUpperCase(),
      if ((countryCode ?? '').trim().isNotEmpty)
        'country_code': countryCode!.trim().toUpperCase(),
      if ((hotelName ?? '').trim().isNotEmpty) 'hotel_name': hotelName!.trim(),
      if ((venueName ?? '').trim().isNotEmpty) 'venue_name': venueName!.trim(),
      if ((eventName ?? '').trim().isNotEmpty) 'event_name': eventName!.trim(),
      if ((city ?? '').trim().isNotEmpty) 'city': city!.trim(),
      if ((postcode ?? '').trim().isNotEmpty) 'postcode': postcode!.trim(),
      'ratehawk_hotel_id': ratehawkHotelId,
      if (manual) 'manual': true,
    };
    return json;
  }

  static LimousineTransferEndpoint? tryParse(Object? raw) {
    if (raw is! Map) return null;
    final map = raw.map((key, value) => MapEntry(key.toString(), value));
    final kind = (map['kind'] ?? '').toString().trim();
    if (!LimousineTransferEndpointKind.all.contains(kind)) return null;
    double? number(Object? value) {
      if (value is num) return value.toDouble();
      return double.tryParse('${value ?? ''}');
    }

    String? text(Object? value) {
      final out = (value ?? '').toString().trim();
      return out.isEmpty ? null : out;
    }

    return LimousineTransferEndpoint(
      kind: kind,
      displayName: text(map['display_name'] ?? map['displayName']) ?? '',
      formattedAddress:
          text(map['formatted_address'] ?? map['formattedAddress']) ?? '',
      latitude: number(map['latitude'] ?? map['lat']),
      longitude: number(map['longitude'] ?? map['lng'] ?? map['lon']),
      providerPlaceId: text(map['provider_place_id'] ?? map['providerPlaceId']),
      airportName: text(map['airport_name'] ?? map['airportName']),
      iataCode: text(map['iata_code'] ?? map['iataCode'])?.toUpperCase(),
      countryCode: text(map['country_code'] ?? map['countryCode'])?.toUpperCase(),
      hotelName: text(map['hotel_name'] ?? map['hotelName']),
      venueName: text(map['venue_name'] ?? map['venueName']),
      eventName: text(map['event_name'] ?? map['eventName']),
      city: text(map['city']),
      postcode: text(map['postcode'] ?? map['postal_code']),
      ratehawkHotelId: text(map['ratehawk_hotel_id'] ?? map['ratehawkHotelId']),
      manual: map['manual'] == true,
    );
  }
}

LimousineTransferEndpoint limousineEndpointFromAddress(LimousineAddressValue value) {
  return LimousineTransferEndpoint(
    kind: LimousineTransferEndpointKind.address,
    displayName: value.routeText,
    formattedAddress: value.routeText,
    latitude: value.lat,
    longitude: value.lon,
    providerPlaceId: value.placeId,
    manual: value.acceptance == LimousineAddressAcceptance.manualFallback,
  );
}

LimousineAddressValue limousineAddressValueFromEndpoint(
  LimousineTransferEndpoint endpoint,
) {
  return LimousineAddressValue(
    displayText: endpoint.routeText,
    canonicalLabel: endpoint.routeText,
    lat: endpoint.latitude,
    lon: endpoint.longitude,
    placeId: endpoint.providerPlaceId,
    acceptance: LimousineAddressAcceptance.selected,
  );
}

LimousineTransferEndpoint limousineEndpointFromAirport(AirportCatalogAirport airport) {
  return LimousineTransferEndpoint(
    kind: LimousineTransferEndpointKind.airport,
    displayName: airport.displayLabel,
    formattedAddress: airport.formattedAddress,
    latitude: airport.latitude,
    longitude: airport.longitude,
    airportName: airport.name,
    iataCode: airport.iata,
    countryCode: airport.countryCode,
    city: airport.city,
  );
}

LimousineTransferEndpoint limousineEndpointRetainAddressOnKindChange(
  LimousineTransferEndpoint current,
  String nextKind,
) {
  return LimousineTransferEndpoint(
    kind: nextKind,
    displayName: current.displayName,
    formattedAddress: current.formattedAddress,
    latitude: current.latitude,
    longitude: current.longitude,
    providerPlaceId: nextKind == LimousineTransferEndpointKind.address
        ? current.providerPlaceId
        : null,
    city: current.city,
    postcode: current.postcode,
    countryCode: current.countryCode,
    manual: nextKind != LimousineTransferEndpointKind.airport,
  );
}

({LimousineTransferEndpoint from, LimousineTransferEndpoint to})
    reverseLimousineEndpoints({
  required LimousineTransferEndpoint from,
  required LimousineTransferEndpoint to,
}) {
  return (from: to, to: from);
}

bool limousineAirportEndpointIsCanonical(
  LimousineTransferEndpoint endpoint, {
  List<AirportCatalogAirport>? airports,
}) {
  if (endpoint.kind != LimousineTransferEndpointKind.airport) return false;
  final iata = (endpoint.iataCode ?? '').trim();
  final country = (endpoint.countryCode ?? '').trim();
  final match = airportByIata(iata, countryCode: country, airports: airports);
  if (match == null) return false;
  final name = (endpoint.airportName ?? '').trim();
  if (name.isNotEmpty && name != match.name) {
    return match.iata == iata && match.countryCode == country;
  }
  return true;
}

bool limousineHotelEndpointIsUsable(LimousineTransferEndpoint endpoint) {
  if (endpoint.kind != LimousineTransferEndpointKind.hotel) return false;
  final name = (endpoint.hotelName ?? endpoint.displayName).trim();
  final address = endpoint.formattedAddress.trim();
  if (name.isEmpty || address.isEmpty) return false;
  if (endpoint.manual) return true;
  return endpoint.hasCoordinates;
}

bool limousineEventEndpointIsUsable(LimousineTransferEndpoint endpoint) {
  if (!LimousineTransferEndpointKind.isEvent(endpoint.kind)) return false;
  final name = (endpoint.venueName ?? endpoint.displayName).trim();
  final address = endpoint.formattedAddress.trim();
  if (name.isEmpty || address.isEmpty) return false;
  if (endpoint.manual) return true;
  return endpoint.hasCoordinates;
}

class LimousineItineraryEndpoints {
  const LimousineItineraryEndpoints({
    this.from,
    this.to,
    this.returnPickup,
    this.returnDestination,
    this.stops = const <LimousineTransferEndpoint>[],
    this.airportDirection = '',
    this.hotelDirection = '',
  });

  final LimousineTransferEndpoint? from;
  final LimousineTransferEndpoint? to;
  final LimousineTransferEndpoint? returnPickup;
  final LimousineTransferEndpoint? returnDestination;
  final List<LimousineTransferEndpoint> stops;
  final String airportDirection;
  final String hotelDirection;

  LimousineItineraryEndpoints copyWith({
    LimousineTransferEndpoint? from,
    LimousineTransferEndpoint? to,
    LimousineTransferEndpoint? returnPickup,
    LimousineTransferEndpoint? returnDestination,
    List<LimousineTransferEndpoint>? stops,
    String? airportDirection,
    String? hotelDirection,
    bool clearFrom = false,
    bool clearTo = false,
  }) {
    return LimousineItineraryEndpoints(
      from: clearFrom ? null : (from ?? this.from),
      to: clearTo ? null : (to ?? this.to),
      returnPickup: returnPickup ?? this.returnPickup,
      returnDestination: returnDestination ?? this.returnDestination,
      stops: stops ?? this.stops,
      airportDirection: airportDirection ?? this.airportDirection,
      hotelDirection: hotelDirection ?? this.hotelDirection,
    );
  }

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      if (from != null && !from!.isEmpty) 'from_endpoint': from!.toJson(),
      if (to != null && !to!.isEmpty) 'to_endpoint': to!.toJson(),
      if (returnPickup != null && !returnPickup!.isEmpty)
        'return_pickup_endpoint': returnPickup!.toJson(),
      if (returnDestination != null && !returnDestination!.isEmpty)
        'return_destination_endpoint': returnDestination!.toJson(),
      if (stops.isNotEmpty)
        'stop_endpoints': stops
            .where((stop) => !stop.isEmpty)
            .map((stop) => stop.toJson())
            .toList(growable: false),
      if (airportDirection.trim().isNotEmpty)
        'airport_direction': airportDirection.trim(),
      if (hotelDirection.trim().isNotEmpty)
        'hotel_direction': hotelDirection.trim(),
    };
  }
}

LimousineItineraryEndpoints applyAirportDirection({
  required String direction,
  required AirportCatalogAirport airport,
  LimousineTransferEndpoint? other,
  LimousineItineraryEndpoints current = const LimousineItineraryEndpoints(),
}) {
  final airportEndpoint = limousineEndpointFromAirport(airport);
  final toAirport = direction == 'to_airport';
  return current.copyWith(
    airportDirection: toAirport ? 'to_airport' : 'from_airport',
    from: toAirport ? other : airportEndpoint,
    to: toAirport ? airportEndpoint : other,
    clearFrom: toAirport && other == null,
    clearTo: !toAirport && other == null,
  );
}

LimousineItineraryEndpoints applyHotelDirection({
  required String direction,
  required LimousineTransferEndpoint hotel,
  LimousineTransferEndpoint? other,
  LimousineItineraryEndpoints current = const LimousineItineraryEndpoints(),
}) {
  final toHotel = direction != 'from_hotel';
  return current.copyWith(
    hotelDirection: toHotel ? 'to_hotel' : 'from_hotel',
    from: toHotel ? other : hotel,
    to: toHotel ? hotel : other,
    clearFrom: toHotel && other == null,
    clearTo: !toHotel && other == null,
  );
}

LimousineItineraryEndpoints clearIncompatibleAirportOnCountryChange({
  required LimousineItineraryEndpoints current,
  required String countryCode,
  List<AirportCatalogAirport>? airports,
}) {
  LimousineTransferEndpoint? keep(LimousineTransferEndpoint? endpoint) {
    if (endpoint == null) return null;
    if (endpoint.kind != LimousineTransferEndpointKind.airport) return endpoint;
    if ((endpoint.countryCode ?? '').toUpperCase() == countryCode.toUpperCase() &&
        limousineAirportEndpointIsCanonical(endpoint, airports: airports)) {
      return endpoint;
    }
    return limousineEndpointRetainAddressOnKindChange(
      endpoint,
      LimousineTransferEndpointKind.address,
    );
  }

  return current.copyWith(
    from: keep(current.from),
    to: keep(current.to),
    returnPickup: keep(current.returnPickup),
    returnDestination: keep(current.returnDestination),
  );
}

File: workers/booking/modules/limousine_transfer_endpoint.mjs
// Additive typed itinerary endpoints for limousine quote/book.
// Does not introduce a second catalog, RateHawk client, or booking aggregate.

const KINDS = Object.freeze(["address", "airport", "hotel", "event", "venue"]);

export function limousineEndpointIsEventKind(kind) {
  return kind === "event" || kind === "venue";
}

function asObject(value) {
  return value && typeof value === "object" && !Array.isArray(value) ? value : {};
}

function safeText(value, max = 240) {
  return String(value ?? "").trim().slice(0, max);
}

function toFinite(value) {
  const n = typeof value === "number" ? value : Number(value);
  return Number.isFinite(n) ? n : null;
}

function endpointFingerprint(endpoint) {
  const e = asObject(endpoint);
  return [
    safeText(e.kind, 16),
    safeText(e.display_name, 160).toLowerCase(),
    safeText(e.formatted_address, 240).toLowerCase(),
    safeText(e.iata_code, 8).toUpperCase(),
    safeText(e.country_code, 8).toUpperCase(),
    safeText(e.hotel_name, 120).toLowerCase(),
    safeText(e.venue_name, 160).toLowerCase(),
    safeText(e.event_name, 160).toLowerCase(),
    safeText(e.provider_place_id, 80),
    e.manual === true ? "manual" : "live",
    e.latitude ?? "",
    e.longitude ?? "",
  ].join("|");
}

export function sanitizeLimousineTransferEndpoint(raw) {
  const src = asObject(raw);
  const kind = safeText(src.kind, 16);
  if (!KINDS.includes(kind)) return null;
  const displayName = safeText(src.display_name ?? src.displayName, 240);
  const formattedAddress = safeText(
    src.formatted_address ?? src.formattedAddress,
    320,
  );
  if (!displayName && !formattedAddress) return null;
  const latitude = toFinite(src.latitude ?? src.lat);
  const longitude = toFinite(src.longitude ?? src.lng ?? src.lon);
  const endpoint = {
    kind,
    display_name: displayName || formattedAddress,
    formatted_address: formattedAddress || displayName,
    latitude,
    longitude,
    provider_place_id: safeText(src.provider_place_id ?? src.providerPlaceId, 120) || null,
    country_code: safeText(src.country_code ?? src.countryCode, 8).toUpperCase() || null,
    city: safeText(src.city, 80) || null,
    postcode: safeText(src.postcode ?? src.postal_code, 16) || null,
    manual: src.manual === true,
  };

  if (kind === "airport") {
    const iata = safeText(src.iata_code ?? src.iataCode, 8).toUpperCase();
    const country = safeText(src.country_code ?? src.countryCode, 8).toUpperCase();
    const airportName = safeText(src.airport_name ?? src.airportName, 160);
    if (!/^[A-Z]{3}$/.test(iata) || !/^[A-Z]{2}$/.test(country)) return null;
    endpoint.airport_name = airportName || displayName;
    endpoint.iata_code = iata;
    endpoint.country_code = country;
    return endpoint;
  }

  if (kind === "hotel") {
    const hotelName = safeText(src.hotel_name ?? src.hotelName, 160) || displayName;
    const address = formattedAddress || displayName;
    if (!hotelName || !address) return null;
    if (src.manual !== true && (latitude == null || longitude == null)) return null;
    endpoint.hotel_name = hotelName;
    endpoint.ratehawk_hotel_id =
      safeText(src.ratehawk_hotel_id ?? src.ratehawkHotelId, 64) || null;
    return endpoint;
  }

  if (limousineEndpointIsEventKind(kind)) {
    const venueName =
      safeText(src.venue_name ?? src.venueName, 160) || displayName;
    const address = formattedAddress || displayName;
    if (!venueName || !address) return null;
    if (src.manual !== true && (latitude == null || longitude == null)) return null;
    endpoint.venue_name = venueName;
    endpoint.event_name = safeText(src.event_name ?? src.eventName, 160) || null;
    return endpoint;
  }

  return endpoint;
}

export function sanitizeLimousineItineraryEndpoints(input) {
  const src = asObject(input);
  const from = sanitizeLimousineTransferEndpoint(src.from_endpoint ?? src.fromEndpoint);
  const to = sanitizeLimousineTransferEndpoint(src.to_endpoint ?? src.toEndpoint);
  const returnPickup = sanitizeLimousineTransferEndpoint(
    src.return_pickup_endpoint ?? src.returnPickupEndpoint,
  );
  const returnDestination = sanitizeLimousineTransferEndpoint(
    src.return_destination_endpoint ?? src.returnDestinationEndpoint,
  );
  const stops = Array.isArray(src.stop_endpoints ?? src.stopEndpoints)
    ? (src.stop_endpoints ?? src.stopEndpoints)
        .map((item) => sanitizeLimousineTransferEndpoint(item))
        .filter(Boolean)
        .slice(0, 8)
    : [];
  return {
    from_endpoint: from,
    to_endpoint: to,
    return_pickup_endpoint: returnPickup,
    return_destination_endpoint: returnDestination,
    stop_endpoints: stops,
    airport_direction: safeText(src.airport_direction ?? src.airportDirection, 24),
    hotel_direction: safeText(src.hotel_direction ?? src.hotelDirection, 24),
    itinerary_endpoint_fingerprint: [
      endpointFingerprint(from),
      endpointFingerprint(to),
      endpointFingerprint(returnPickup),
      endpointFingerprint(returnDestination),
    ].join("~"),
  };
}

export function deriveLimousineAirportPricingFacts(input) {
  const src = asObject(input);
  const itinerary = sanitizeLimousineItineraryEndpoints(src);
  const from = itinerary.from_endpoint;
  const to = itinerary.to_endpoint;
  let airportIata = safeText(src.airport_iata ?? src.airportIata, 8).toUpperCase();
  let direction = safeText(
    src.airport_direction ?? src.direction ?? itinerary.airport_direction,
    24,
  );
  if (to?.kind === "airport" && to.iata_code) {
    if (!airportIata) airportIata = String(to.iata_code).toUpperCase();
    if (!direction) direction = "to_airport";
  } else if (from?.kind === "airport" && from.iata_code) {
    if (!airportIata) airportIata = String(from.iata_code).toUpperCase();
    if (!direction) direction = "from_airport";
  }
  return {
    airport_iata: airportIata,
    direction,
    itinerary,
  };
}

export function attachLimousineItineraryEndpoints(target, input) {
  const next = { ...asObject(target) };
  const itinerary = sanitizeLimousineItineraryEndpoints(input);
  if (itinerary.from_endpoint) next.from_endpoint = itinerary.from_endpoint;
  if (itinerary.to_endpoint) next.to_endpoint = itinerary.to_endpoint;
  if (itinerary.return_pickup_endpoint) {
    next.return_pickup_endpoint = itinerary.return_pickup_endpoint;
  }
  if (itinerary.return_destination_endpoint) {
    next.return_destination_endpoint = itinerary.return_destination_endpoint;
  }
  if (itinerary.stop_endpoints.length) next.stop_endpoints = itinerary.stop_endpoints;
  if (itinerary.airport_direction) next.airport_direction = itinerary.airport_direction;
  if (itinerary.hotel_direction) next.hotel_direction = itinerary.hotel_direction;
  next.itinerary_endpoint_fingerprint = itinerary.itinerary_endpoint_fingerprint;
  return next;
}

export function limousineItineraryConflictsWithJourney(journeyType, itinerary) {
  const journey = String(journeyType ?? "").trim();
  const src = asObject(itinerary);
  if (journey === "event_transfer" && src.to_endpoint) {
    return !limousineEndpointIsEventKind(src.to_endpoint.kind);
  }
  return false;
}
